File: bin/asset.py
# Copyright 2014 Tom SF Haines

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import xml.sax.saxutils as saxutils

from .login import app as login
from .misc import format_bytes



def app(rfam, response):
  # If the user is not logged in show them the login page...
  if response.user==None or response.project==None:
    login(rfam, response)
    return
  
  # Get the information for the requested asset...
  db = rfam.proj(response.project)
  json_path = response.getPath()[1:]
  path = '/'.join(json_path)
  json_path = json_path[:-1] + [json_path[-1] + '.json']
  asset = db[json_path].read()
  
  try:
    blend = db[response.getPath()[1:]]
  except KeyError:
    blend = None
  
  # Prepare the back/new buttons...
  new_button = rfam.template('button.back_new', {}, response)
  
  # Prepare assorted bits of information...
  type_choice = rfam.typeChoice(response.project, asset['type'])
  state = rfam.stateChoice(response.project, asset['type'], asset['state'])
  priority = rfam.priorityInterface(response.project, asset['priority'])
  render = 'checked' if asset['render'] else ''
  final = 'checked' if asset['final'] else ''
  
  owner = rfam.userChoice(response.project, asset['owner'], True)
  
  support = []
  
  for ident in rfam.getUsersByProject(response.project):
    user = rfam.getUser(ident)
    checked = 'checked' if ident in asset['support'] else ''
    support.append(rfam.template('asset.support', {'ident':ident, 'name' : saxutils.escape(user['name']), 'checked' : checked}, response))
  
  support = '\n'.join(support)
  
  render_stats = rfam.format_render_time(response.user, asset)
  
  modified = blend.modified().strftime(rfam.getLanguage(response.user)['asset:modified:format']) if blend!=None else rfam.getLanguage(response.user)['NA']
  size = format_bytes(blend.size()) if blend!=None else rfam.getLanguage(response.user)['NA']
  if asset['creator']!=None:
    creator = rfam.getUser(asset['creator'])['name']
  else:
    creator = rfam.getLanguage(response.user)['asset:unknown_user']
  
  if blend==None:
    path_extra = rfam.getLanguage(response.user)['not_found']
  else:
    path_extra = None
  
  # Show the asset page...
  head = '<link rel="stylesheet" href="/stylesheets/asset.css"/>\n<script src="/javascript/asset.js"></script>'
  payload = {'title' : response.getPath()[-1], 'header_extra' : new_button, 'head' : head, 'path' : path, 'path_extra' : path_extra, 'name' : asset['name'], 'description' : asset['description'], 'type' : type_choice, 'owner' : owner, 'support' : support, 'state' : state, 'priority' : priority, 'render' : render, 'final' : final, 'render_stats' : render_stats, 'modified' : modified, 'size' : size, 'creator' : creator}
  html = rfam.template('asset', payload, response)
  response.append(html)
  response.setHTML()

File: javascript/asset.js
// Copyright 2014 Tom SF Haines

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.



// Button click handlers...
function reset_stats()
{
 // Extract the path of the asset...
  var path = $('#page_tab').attr('data-path');
  
 // Callback of response - refresh the page...
  var callback = function(response)
  {
   if (response)
   {
    location.reload();
   }
   else
   {
    alert('Error reseting render statistics');  
   }
  };
  
 // Send the request...
  $.getJSON('/action/reset_stats/' + path, {}, callback);
}


function checkpoint()
{
 // Callback of response - refresh the page...
  var callback = function(response)
  {
   if (response)
   {
    var d = new Date();
    $('#checkpoint_result').html('Checkpointed at ' + d.toTimeString());
   }
   else
   {
    $('#checkpoint_result').html('Error: Checkpoint not created');
   }
  };
  
 // Send the request...
  path = $('#page_tab').attr('data-path');
  $.getJSON('/action/checkpoint/' + path, {}, callback);
}

function delete_asset()
{
 if (confirm('Are you sure you want to delete this asset?'))
 {
  // Callback of response - refresh the page...
   var callback = function(response)
   {
    if (response)
    {
     window.location.pathname = '/assets';
    }
    else
    {
     alert('Failed to delete file');
    }
   };
  
  // Send the request...
   path = $('#page_tab').attr('data-path');
   $.getJSON('/action/delete/' + path, {}, callback);
 }
}



// Setup all the handlers...
$(function() {
 // Extract the path of the asset...
  var path = $('#page_tab').attr('data-path');
  
 // Do the many handlers...
  tie_input('#name', 'asset/' + path, 'name');
  tie_input('#description', 'asset/' + path, 'description');
  
  tie_input('#type', 'asset/' + path, 'type');
  tie_input('#state', 'asset/' + path, 'state');
  tie_input('#priority', 'asset/' + path, 'priority');
  tie_checkbox('#render', 'asset/' + path, 'render');
  tie_checkbox('#final', 'asset/' + path, 'final');
  
  tie_input('#owner', 'asset/' + path, 'owner');
  tie_checkbox('.supporter', 'asset/' + path, 'support');
  
 // Make the back button a little smarter...
  $('#back').click(function()
  {
   var split = window.location.href.split('?');
   if (split.length>1)
   {
    dest = split[1];
    if (dest=='home')
    {
     window.location.pathname = '/home';
     return false;
    }
    if (dest=='team')
    {
     window.location.pathname = '/team';
     return false;
    }
   }
  });
  
 // Setup button handlers...
  $('#reset_stats').click(reset_stats);
  $('#checkpoint').click(checkpoint);
  $('#delete').click(delete_asset);
});
